### PR TITLE
[SPARK-38444][SQL]Automatically calculate the upper and lower bounds of partitions when no specified partition related params

### PR DESCRIPTION
**Pull request by caican00**
### What changes were proposed in this pull request?
when access rdbms, such as mysql, this patch can automatically calculate upper and lower bounds according to the primary key to improve parallelism and speed up query.

### Why are the changes needed?
when access rdbms, such as mysql, if partitionColumn, lowerBound, upperBound, numPartitions are not specified, by default only one partition to scan database is working. 

It makes load data from database slow and makes it difficult for users to configure multiple parameters to improve parallelism.

This patch can automatically calculate upper and lower bounds according to the primary key to improve parallelism and speed up query.

### Does this PR introduce _any_ user-facing change?
yes. new config `defaultNumPartitions` in JDBCOptions. It is used to set the default parallelism.

### How was this patch tested?
new testing

### CODE REVIEW

**Comment by caican00** [comment]
@MaxGekk  HI, could you help to review this patch? thanks

**Comment by HyukjinKwon** [comment]
cc @maropu too FYI who looked into these code paths.

**Comment by caican00** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala [hunk 1]
```diff
@@ -111,6 +111,9 @@ class JDBCOptions(
   // the number of partitions
   val numPartitions = parameters.get(JDBC_NUM_PARTITIONS).map(_.toInt)
 
+  // the default number of partitions
+  val defaultNumPartitions = parameters.getOrElse(DEFAULT_NUM_PARTITIONS, "10").toInt
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala:115]: What is the difference between `JDBC_NUM_PARTITIONS`?

Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala:115]: Same question here :)
Do we really need a default? If the users want to have multiple partitions, shouldn't they specify this explicitly?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 2]
```diff
@@ -168,6 +177,71 @@ private[sql] object JDBCRelation extends Logging {
...
+   * get the min and max value by the column
+   * @param schema resolved schema of a JDBC table
+   * @param resolver function used to determine if two identifiers are equal
+   * @param timeZoneId timezone ID to be used if a partition column type is date or timestamp
+   * @param jdbcOptions JDBC options that contains url
+   * @param filters filters in Where clause
+   * @return JDBCPartitioningInfo
+   */
+  def getPartitionBound(
+                         schema: StructType,
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala:190]: nit: 4 space indentation

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalogSuite.scala [hunk 3]
```diff
@@ -425,4 +430,48 @@ class JDBCTableCatalogSuite extends QueryTest with SharedSparkSession {
...
+    // scan empty table
+    val empty_df = sql("select id, name from h2.test.employee")
+    // default partition num is 10
+    assert(empty_df.rdd.getNumPartitions == 1)
+    sql("insert overwrite h2.test.employee select 1000 as id, 'a' as name")
+    for (id <- 1 to 100) {
+      sql(s"insert into h2.test.employee values($id, 'a')")
+    }
+    val df = sql("select id, name from h2.test.employee")
+    // default partition num is 15
```
Comment (huaxingao) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalogSuite.scala:450]: you mean `default partition num is 10`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala (hunk 1 above)
Comment (caican00) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala:115]: > What is the difference between `JDBC_NUM_PARTITIONS`?

@wangyum 
1. partitionColumn, lowerBound, upperBound and numPartitions must be specified together. If an unreasonable numPartitions was specified by users, such as 1, the parallelism is still very small.
2. therefore, we(jdbc) should rezoning partition nums using another config, not JDBC_NUM_PARTITIONS(it is specified by users and its value maybe very small)

Comment (caican00) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala:115]: > Same question here :) Do we really need a default? If the users want to have multiple partitions, shouldn't they specify this explicitly?

@huaxingao 
1. if users want to change partition nums, they only need to specify `DEFAULT_NUM_PARTITIONS` explicitly and do not need to specify any more parameters, jdbc will automatically calculate the upper and lower bounds of partitions using primary key
2. If users do not specify any parameters, then we(jdbc) need a default value to determine the base number of partitions

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala (hunk 2 above)
Comment (caican00) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala:190]: thanks, updated it.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalogSuite.scala (hunk 3 above)
Comment (caican00) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalogSuite.scala]: my mistake. updated it, thanks

**Comment by cloud-fan** [comment]
@sadikovi can you take a look?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 4]
```diff
@@ -168,6 +177,71 @@ private[sql] object JDBCRelation extends Logging {
     partitions
   }
 
+  /**
+   * get the min and max value by the column
```
Comment (sadikovi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala:181]: nit: Get the min and max values for the column.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 5]
```diff
@@ -168,6 +177,71 @@ private[sql] object JDBCRelation extends Logging {
...
+   * @param jdbcOptions JDBC options that contains url
+   * @param filters filters in Where clause
+   * @return JDBCPartitioningInfo
+   */
+  def getPartitionBound(
+      schema: StructType,
+      resolver: Resolver,
+      timeZoneId: String,
+      jdbcOptions: JDBCOptions,
+      filters: Array[Filter] = Array.empty): JDBCPartitioningInfo = {
```
Comment (sadikovi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala:194]: Shall we return `Option[JDBCPartitioningInfo]` instead?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 6]
```diff
@@ -168,6 +177,71 @@ private[sql] object JDBCRelation extends Logging {
...
+        (f.dataType.isInstanceOf[NumericType] ||
+          f.dataType.isInstanceOf[DateType] ||
+          f.dataType.isInstanceOf[TimestampType]))
+
+    if (prks.length > 0) {
+      val prk = prks.head
+      val dataType = prk.dataType
+      var lBound: String = null
+      var uBound: String = null
+      val sql = s"select min(${prk.name}) as lBound, max(${prk.name}) as uBound " +
```
Comment (sadikovi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala:211]: Can you explain this logic in the javadoc for this method? Also, what happens if the table is empty?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 7]
```diff
@@ -168,6 +177,71 @@ private[sql] object JDBCRelation extends Logging {
...
+        s"from ${jdbcOptions.tableOrQuery} limit 1"
+      val conn = JdbcDialects.get(jdbcOptions.url).createConnectionFactory(jdbcOptions)(-1)
+      try {
+        val statement = conn.prepareStatement(sql)
+        try {
+          statement.setQueryTimeout(jdbcOptions.queryTimeout)
+          val resultSet = statement.executeQuery()
+          while (resultSet.next()) {
+            lBound = resultSet.getString("lBound")
```
Comment (sadikovi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala:220]: Would it work for primary keys that are integers or timestamps?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 8]
```diff
@@ -168,6 +177,71 @@ private[sql] object JDBCRelation extends Logging {
...
+            uBound = resultSet.getString("uBound")
+          }
+        } catch {
+          case _: SQLException =>
```
Comment (sadikovi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala:224]: Maybe it is worth at least logging the exception but I would consider re-throwing it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 9]
```diff
@@ -168,6 +177,71 @@ private[sql] object JDBCRelation extends Logging {
...
+      resolver: Resolver,
+      timeZoneId: String,
+      jdbcOptions: JDBCOptions,
+      filters: Array[Filter] = Array.empty): JDBCPartitioningInfo = {
+    // columns in filters
+    val filterColumns = new util.ArrayList[String]()
+    filters.map(filter => filter.references.distinct.map(r => filterColumns.add(r)))
+    // primary keys used for partitioning
+    val prks = schema.fields.filter(
+      f => f.metadata.getBoolean("isIndexKey") &&
```
Comment (sadikovi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala:200]: Does the code handle composite primary keys or any multi-column indexes, e.g. with 2 or more columns?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 10]
```diff
@@ -306,6 +318,12 @@ object JdbcUtils extends Logging with SQLConfHelper {
       val columnType =
         dialect.getCatalystType(dataType, typeName, fieldSize, metadata).getOrElse(
           getCatalystType(dataType, fieldSize, fieldScale, isSigned))
+      list.contains(columnName) match {
```
Comment (sadikovi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:321]: Is it the same as:
```scala
metadata.putBoolean("isIndexKey", list.contains(columnName))
```

Also, can we make list a set?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala [hunk 11]
```diff
@@ -111,6 +111,9 @@ class JDBCOptions(
   // the number of partitions
   val numPartitions = parameters.get(JDBC_NUM_PARTITIONS).map(_.toInt)
 
+  // the default number of partitions
```
Comment (sadikovi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala:114]: Can you update this comment? It is unclear what default number of partition it is - is it for overall number of partitions in the RDD or is it specifically for primary keys in the table and pushed filters?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala (hunk 1 above)
Comment (sadikovi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala:115]: I think the name of the config is misleading, this is essentially the default value of `numPartitions` configs:
```scala
val numPartitions = parameters.get(JDBC_NUM_PARTITIONS).map(_.toInt).getOrElse(10)
```

**Review by sadikovi — COMMENTED** [review]
Can you update the PR title and description to reflect the changes? I think we should just have a flag to enable/disable partitioning based on available primary keys.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala (hunk 1 above)
Comment (sadikovi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala:115]: I replied below. I think if this is about default number of partitions, we can just set the default to JDBC_NUM_PARTITIONS flag.

**Pull request closed without merging**